server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeRole.java

```diff
@@ -242,4 +261,84 @@ public void validateRoles(final List<DiscoveryNodeRole> roles) {
+            .filter(f -> f.getType().equals(DiscoveryNodeRole.class))
+            .collect(Collectors.toUnmodifiableList());
+        // this will detect duplicate role names
+        final Map<String, DiscoveryNodeRole> roleMap = roleFields.stream()
+            .map(f -> {
+                try {
+                    return (DiscoveryNodeRole) f.get(null);
+                } catch (final IllegalAccessException e) {
+                    throw new AssertionError(e);
+                }
+            })
+            .collect(Collectors.toMap(DiscoveryNodeRole::roleName, Function.identity()));
+        assert roleMap.size() == roleFields.size() :
+            "roles by name [" + roleMap + "], role fields [" + roleFields + "]";
+        final SortedSet<DiscoveryNodeRole> roles = roleMap.values().stream().collect(Sets.toUnmodifiableSortedSet());
+        // collect the abbreviation names into a map to ensure that there are not any duplicate abbreviations
+        final Map<String, DiscoveryNodeRole> abbreviations = roles
+            .stream()
+            .collect(Collectors.toUnmodifiableMap(DiscoveryNodeRole::roleNameAbbreviation, Function.identity()));
+        assert roleMap.size() == abbreviations.size() :
+            "roles by name [" + roleMap + "], roles by name abbreviation [" + abbreviations + "]";
+        ROLES = roles;
+        ROLE_MAP = Map.copyOf(roleMap); // this ensures ROLE_MAP is immutable
+    }
+
+    /**
+     * The set of possible role names.
+     *
+     * @return an ordered, immutable set of possible role names
+     */
+    public static SortedSet<String> roleNames() {
+        return ROLE_MAP.keySet().stream().collect(Sets.toUnmodifiableSortedSet());
+    }
+
+    /**
+     * Get an optional representing the role with the given role name, if such a role exists.
+     *
+     * @param roleName the role name to get the associated role representation for
+     * @return an optional node role
+     */
+    public static Optional<DiscoveryNodeRole> maybeGetRoleFromRoleName(final String roleName) {
+        if (ROLE_MAP.containsKey(roleName) == false) {
+            return Optional.empty();
+        }
+        return Optional.of(ROLE_MAP.get(roleName));
+    }
+
+    /**
+     * Get a representation of the role with the given role name, if such a role exists, otherwise an exception is thrown.
+     *
+     * @param roleName the role name to get the associated role representation for
+     * @return a node role
+     *
+     * @throws IllegalArgumentException if no node role with the given role name exists
+     */
+    public static DiscoveryNodeRole getRoleFromRoleName(final String roleName) {
+        final Optional<DiscoveryNodeRole> maybeRole = maybeGetRoleFromRoleName(roleName);
+        if (maybeRole.isEmpty()) {
+            throw new IllegalArgumentException("unknown role [" + roleName + "]");
+        }
+        return maybeRole.get();
```

Comment:
As long as we're code golfing
```suggestion
        return maybeGetRoleFromRoleName(roleName)
            .orElseThrow(() -> new IllegalArgumentException("unknown role [" + roleName + "]));
```

Reply:
Thanks.